### PR TITLE
pageserver: issue concurrent IO on the read path

### PR DESCRIPTION
**Pull request by VladLazar**
## Refs

- Epic: https://github.com/neondatabase/neon/issues/9378

Co-authored-by: Vlad Lazar <vlad@neon.tech>
Co-authored-by: Christian Schwarz <christian@neon.tech>

## Problem

The read path does its IOs sequentially.
This means that if N values need to be read to reconstruct a page,
we will do N IOs and getpage latency is `O(N*IoLatency)`.

## Solution

With this PR we gain the ability to issue IO concurrently within one
layer visit **and** to move on to the next layer without waiting for IOs
from the previous visit to complete.

This is an evolved version of the work done at the Lisbon hackathon,
cf https://github.com/neondatabase/neon/pull/9002.

### `will_init` now sourced from disk btree index keys

On the algorithmic level, the only change is that the `get_values_reconstruct_data`
now sources `will_init` from the disk btree index key (which is PS-page_cache'd), instead
of from the `Value`, which is only available after the IO completes.

### Concurrent IOs, Submission & Completion 

To separate IO submission from waiting for its completion, while simultaneously
feature-gating the change, we introduce the notion of an `IoConcurrency` struct
through which IO futures are "spawned".

An IO is an opaque future, and waiting for completions is handled through
`tokio::sync::oneshot` channels.
The oneshot Receiver's take the place of the `img` and `records` fields
inside `VectoredValueReconstructState`.

When we're done visiting all the layers and submitting all the IOs along the way
we concurrently `collect_pending_ios` for each value, which means
for each value there is a future that awaits all the oneshot receivers
and then calls into walredo to reconstruct the page image.
Walredo is now invoked concurrently for each value instead of sequentially.
Walredo itself remains unchanged.

The spawned IO futures are driven to completion by a sidecar tokio task that
is separate from the task that performs all the layer visiting and spawning of IOs.
That tasks receives the IO futures via an unbounded mpsc channel and
drives them to completion inside a `FuturedUnordered`.

(The behavior from before this PR is available through `IoConcurrency::Sequential`,
which awaits the IO futures in place, without "spawning" or "submitting" them
anywhere.)

#### Alternatives Explored

A few words on the rationale behind having a sidecar *task* and what
alternatives were considered.

One option is to queue up all IO futures in a FuturesUnordered that is polled
the first time when we `collect_pending_ios`.

Firstly, the IO futures are opaque, compiler-generated futures that need
to be polled at least once to submit their IO. "At least once" because
tokio-epoll-uring may not be able to submit the IO to the kernel on first
poll right away.

Second, there are deadlocks if we don't drive the IO futures to completion
independently of the spawning task.
The reason is that both the IO futures and the spawning task may hold some
_and_ try to acquire _more_ shared limited resources.
For example, both spawning task and IO future may try to acquire
* a VirtualFile file descriptor cache slot async mutex (observed during impl)
* a tokio-epoll-uring submission slot (observed during impl)
* a PageCache slot (currently this is not the case but we may move more code into the IO futures in the future)

Another option is to spawn a short-lived `tokio::task` for each IO future.
We implemented and benchmarked it during development, but found little
throughput improvement and moderate mean & tail latency degradation.
Concerns about pressure on the tokio scheduler made us discard this variant.

The sidecar task could be obsoleted if the IOs were not arbitrary code but a well-defined struct.
However,
1. the opaque futures approach taken in this PR allows leaving the existing
   code unchanged, which
2. allows us to implement the `IoConcurrency::Sequential` mode for feature-gating
   the change.

Once the new mode sidecar task implementation is rolled out everywhere,
and `::Sequential` removed, we can think about a descriptive submission & completion interface.
The problems around deadlocks pointed out earlier will need to be solved then.
For example, we could eliminate VirtualFile file descriptor cache and tokio-epoll-uring slots.
The latter has been drafted in https://github.com/neondatabase/tokio-epoll-uring/pull/63.

See the lengthy doc comment on `spawn_io()` for more details.

### Error handling

There are two error classes during reconstruct data retrieval:
* traversal errors: index lookup, move to next layer, and the like
* value read IO errors

A traversal error fails the entire get_vectored request, as before this PR.
A value read error only fails that value.

In any case, we preserve the existing behavior that once
`get_vectored` returns, all IOs are done. Panics and failing
to poll `get_vectored` to completion will leave the IOs dangling,
which is safe but shouldn't happen, and so, a rate-limited
log statement will be emitted at warning level.
There is a doc comment on `collect_pending_ios` giving more code-level
details and rationale.

### Feature Gating

The new behavior is opt-in via pageserver config.
The `Sequential` mode is the default.
The only significant change in `Sequential` mode compared to before
this PR is the buffering of results in the `oneshot`s.

## Code-Level Changes

Prep work:
  * Make `GateGuard` clonable.

Core Feature:
* Traversal code: track  `will_init` in `BlobMeta` and source it from
  the Delta/Image/InMemory layer index, instead of determining `will_init`
  after we've read the value. This avoids having to read the value to
  determine whether traversal can stop.
* Introduce `IoConcurrency` & its sidecar task.
  * `IoConcurrency` is the clonable handle.
  * It connects to the sidecar task via an `mpsc`.
* Plumb through `IoConcurrency` from high level code to the
  individual layer implementations' `get_values_reconstruct_data`.
  We piggy-back on the `ValuesReconstructState` for this.
   * The sidecar task should be long-lived, so, `IoConcurrency` needs
     to be rooted up "high" in the call stack.
   * Roots as of this PR:
     * `page_service`: outside of pagestream loop
     * `create_image_layers`: when it is called
     * `basebackup`(only auxfiles + replorigin + SLRU segments)
   * Code with no roots that uses `IoConcurrency::sequential`
     * any `Timeline::get` call
       * `collect_keyspace` is a good example
       * follow-up: https://github.com/neondatabase/neon/issues/10460
     * `TimelineAdaptor` code used by the compaction simulator, unused in practive
     * `ingest_xlog_dbase_create`
* Transform Delta/Image/InMemoryLayer to
  * do their values IO in a distinct `async {}` block
  * extend the residence of the Delta/Image layer until the IO is done
  * buffer their results in a `oneshot` channel instead of straight
    in `ValuesReconstructState` 
  * the `oneshot` channel is wrapped in `OnDiskValueIo` / `OnDiskValueIoWaiter`
    types that aid in expressiveness and are used to keep track of
    in-flight IOs so we can print warnings if we leave them dangling.
* Change `ValuesReconstructState` to hold the receiving end of the
 `oneshot` channel aka `OnDiskValueIoWaiter`.
* Change `get_vectored_impl` to `collect_pending_ios` and issue walredo concurrently, in a `FuturesUnordered`.

Testing / Benchmarking:
* Support queue-depth in pagebench for manual benchmarkinng.
* Add test suite support for setting concurrency mode ps config
   field via a) an env var and b) via NeonEnvBuilder.
* Hacky helper to have sidecar-based IoConcurrency in tests.
   This will be cleaned up later.

More benchmarking will happen post-merge in nightly benchmarks, plus in staging/pre-prod.
Some intermediate helpers for manual benchmarking have been preserved in https://github.com/neondatabase/neon/pull/10466 and will be landed in later PRs.
(L0 layer stack generator!)

Drive-By:
* test suite actually didn't enable batching by default because
  `config.compatibility_neon_binpath` is always Truthy in our CI environment
  => https://neondb.slack.com/archives/C059ZC138NR/p1737490501941309
* initial logical size calculation wasn't always polled to completion, which was
  surfaced through the added WARN logs emitted when dropping a 
  `ValuesReconstructState` that still has inflight IOs.
* remove the timing histograms `pageserver_getpage_get_reconstruct_data_seconds`
  and `pageserver_getpage_reconstruct_seconds` because with planning, value read
  IO, and walredo happening concurrently, one can no longer attribute latency
  to any one of them; we'll revisit this when Vlad's work on tracing/sampling
  through RequestContext lands.
* remove code related to `get_cached_lsn()`.
  The logic around this has been dead at runtime for a long time,
  ever since the removal of the materialized page cache in #8105.

## Testing

Unit tests use the sidecar task by default and run both modes in CI.
Python regression tests and benchmarks also use the sidecar task by default.
We'll test more in staging and possibly preprod.

# Future Work

Please refer to the parent epic for the full plan.

The next step will be to fold the plumbing of IoConcurrency
into RequestContext so that the function signatures get cleaned up.

Once `Sequential` isn't used anymore, we can take the next
big leap which is replacing the opaque IOs with structs
that have well-defined semantics.

### CODE REVIEW

**Comment by erikgrinaker** [comment]
> As of this commit, one IO failure does not stop any other IO requests. When awaiting for the IOs to complete, we stop waiting on the first failure, but we do not signal any other pending IOs to complete and they will just fail silently.

Is this true? It really depends on how the IO futures are implemented, but in general, dropping a future should cancel the in-flight operation and stop polling it. Assuming they're implemented that way, it should be sufficient to ensure that the caller receives the error as soon as it happens and then drops the in-flight futures by returning the error. I don't think we need any synchronization beyond that, or am I missing something?

File: pageserver/src/tenant/storage_layer.rs [hunk 1]
```diff
@@ -200,56 +253,24 @@ impl ValuesReconstructState {
...
-                    }
-                },
-            };
+        let state = self.keys.entry(*key).or_default();
 
-            if key_done && state.situation == ValueReconstructSituation::Continue {
-                state.situation = ValueReconstructSituation::Complete;
-                self.keys_done.add_key(*key);
+        match state.situation {
+            ValueReconstructSituation::Complete => unreachable!(),
```
Comment (erikgrinaker) [inline pageserver/src/tenant/storage_layer.rs:262]: nit: we can turn this into an `assert_eq!` and drop the condition for `push` and in the `completes` branch below.

**Review by erikgrinaker — COMMENTED** [review]
Still reviewing, just flushing some comments for now. All nits, take them or leave them.

File: pageserver/src/tenant/storage_layer.rs [hunk 2]
```diff
@@ -129,17 +193,29 @@ impl ValuesReconstructState {
             keys_with_image_coverage: None,
             layers_visited: 0,
             delta_layers_visited: 0,
+            io_concurrency: {
+                static IO_CONCURRENCY: once_cell::sync::Lazy<String> =
+                    once_cell::sync::Lazy::new(|| {
+                        std::env::var("NEON_PAGESERVER_VALUE_RECONSTRUCT_IO_CONCURRENCY")
+                            .unwrap_or_else(|_| "serial".to_string())
+                    });
```
Comment (erikgrinaker) [inline pageserver/src/tenant/storage_layer.rs:201]: Ref. our discussion last week, it would be nice if we could randomly vary this in tests -- it would give us a ton of test coverage for ~free.

It seems like we don't currently have any test coverage at all for `Parallel`? Let's make sure we have at least some coverage -- either by randomly enabling it, or manually doing so in a few appropriate tests.

File: pageserver/src/tenant/storage_layer.rs [hunk 3]
```diff
@@ -8,6 +8,8 @@ pub(crate) mod layer;
 mod layer_desc;
 mod layer_name;
 pub mod merge_iterator;
+use tokio::sync::{self};
+use utils::bin_ser::BeSer;
```
Comment (erikgrinaker) [inline pageserver/src/tenant/storage_layer.rs:12]: nit: let's move this down to the use blocks below.

File: pageserver/src/tenant/storage_layer/delta_layer.rs [hunk 4]
```diff
@@ -880,8 +878,6 @@ impl DeltaLayerInner {
         self.do_reads_and_update_state(reads, reconstruct_state, ctx)
             .await;
 
-        reconstruct_state.on_lsn_advanced(&keyspace, lsn_range.start);
```
Comment (erikgrinaker) [inline pageserver/src/tenant/storage_layer/delta_layer.rs:883]: Am I understanding it correctly that we can no longer apply this optimization because we're dispatching all the IOs concurrently and we can't know which delta layer reads can be omitted until we've read the image layers?

I guess that's reasonable, unless we want to be clever and use read barriers at each image layer or some other heuristic (not clear that this would be a net win anyway). But what about the default `Serial` path, will that now see a performance regression because we removed this?

File: pageserver/src/tenant/storage_layer/delta_layer.rs [hunk 5]
```diff
@@ -994,94 +988,76 @@ impl DeltaLayerInner {
...
+            > = Default::default();
+            for (_, blob_meta) in read.blobs_at.as_slice().iter().rev() {
+                let (tx, rx) = sync::oneshot::channel();
+                senders.insert((blob_meta.key, blob_meta.lsn), tx);
+                reconstruct_state.update_key(
+                    &blob_meta.key,
+                    blob_meta.lsn,
+                    blob_meta.will_init,
+                    rx,
+                );
```
[the remark this answers is not in the document]
Comment (VladLazar) [reply pageserver/src/tenant/storage_layer/delta_layer.rs:1016]: > Is there a compelling reason why we chose to use channels here instead of making update_key an async function and passing it an IO future for the value?

`update_key` shouldn't wait on the completion of the IO, so I don't really understand why we would make it async.
The receiver is just pushed into `VectoredValueReconstructState::on_disk_values` and that vector is drained at the end of the read in `VectoredValueReconstructState::collect_pending_ios`.

I guess we could change `VectoredValueReconstructState::on_disk_values` to stash IO futures instead of oneshot channels, but I don't really see what that gets us.

File: pageserver/src/tenant/storage_layer/delta_layer.rs (hunk 4 above)
Comment (VladLazar) [reply pageserver/src/tenant/storage_layer/delta_layer.rs:883]: We used to store page images in the pageserver page cache when the vectored read path was implemented. Hence, I also implemented support for using the cache on the vectored read path (that's what this function was part of). Nowadays, we no longer store page images in the page cache (only laryer indices are cached in-mem).

TLDR: I don't expect any perf diff since this wasn't used.

File: pageserver/src/tenant/storage_layer.rs (hunk 2 above)
Comment (VladLazar) [reply pageserver/src/tenant/storage_layer.rs:213]: I've tweaked a unit test in https://github.com/neondatabase/neon/pull/9353/commits/dba696885d151c444e20620db3af3d5b15468c99 to get at least some coverage.

File: pageserver/src/tenant/storage_layer.rs [hunk 6]
```diff
@@ -119,6 +152,37 @@ pub(crate) struct ValuesReconstructState {
...
+
+    io_concurrency: IoConcurrency,
+}
+
+enum IoConcurrency {
+    Serial {
+        prev_io: Option<tokio::task::JoinHandle<()>>,
+    },
+    Parallel,
+}
```
[the remark this answers is not in the document]
Comment (VladLazar) [reply pageserver/src/tenant/storage_layer.rs:169]: https://github.com/neondatabase/neon/pull/9353/commits/2508c64646b1721b6bd3fd8179d7ea02835c226c

File: pageserver/src/tenant/timeline.rs [hunk 7]
```diff
@@ -1133,22 +1134,37 @@ impl Timeline {
...
-                    let state = ValueReconstructState::from(state);
+        let futs = FuturesUnordered::new();
+        for (key, state) in std::mem::take(&mut reconstruct_state.keys) {
+            futs.push({
+                let walredo_self = self.myself.upgrade().expect("&self method holds the arc");
+                async move {
+                    assert!(matches!(
+                        state.situation,
+                        ValueReconstructSituation::Complete
+                    ));
```
[the remark this answers is not in the document]
Comment (VladLazar) [reply pageserver/src/tenant/timeline.rs]: https://github.com/neondatabase/neon/pull/9353/commits/9f95864c8a0cd57205809337bef22971a9578a10

File: pageserver/src/tenant/storage_layer.rs (hunk 1 above)
Comment (VladLazar) [reply pageserver/src/tenant/storage_layer.rs:289]: I'd rather keep it as is. I think the condition below reads easier this way (don't have to figure out control flow).

Comment (erikgrinaker) [reply pageserver/src/tenant/storage_layer.rs:289]: Ok. I find it more confusing myself -- the `unreachable` here basically states that it is illegal to call this method on a `Complete` key, but then the condition below indicates that it is valid. If it's illegal to call this on `Complete` it seems more clear to state that via an assertion at the method entry point.

End result is the same though, so I won't press for it.

File: pageserver/src/tenant/storage_layer/delta_layer.rs (hunk 4 above)
Comment (erikgrinaker) [reply pageserver/src/tenant/storage_layer/delta_layer.rs:883]: Got it, thanks for the background.

File: pageserver/src/tenant/storage_layer.rs (hunk 3 above)
Comment (erikgrinaker) [reply pageserver/src/tenant/storage_layer.rs]: Just that these `use` statements were in the middle of a `mod` block. Seems fixed now.

**Comment by VladLazar** [comment]
> > As of this commit, one IO failure does not stop any other IO requests. When awaiting for the IOs to complete, we stop waiting on the first failure, but we do not signal any other pending IOs to complete and they will just fail silently.
> 
> Is this true? It really depends on how the IO futures are implemented, but in general, dropping a future should cancel the in-flight operation and stop polling it. Assuming they're implemented that way, it should be sufficient to ensure that the caller receives the error as soon as it happens and then drops the in-flight futures by returning the error. I don't think we need any synchronization beyond that, or am I missing something?

I missed this question.

All "IO futures" are collected in `VectoredValueReconstructState::on_disk_values`. With this PR, the read path does not wait for the outcome of one IO before issuing the next one. Hence, at the end of the layer traversal, we will have created "IO futures" for all required values. Each "IO future" is independent. If the first one fails, the others are still present.

Let's also consider what happens when all the "IO futures" after the failed one are not complete. We bail out in `collect_pending_ios`, but the tasks for all the incomplete IOs still run since we don't call `abort` on the `JoinHandle` or have cancellation wired in.

File: pageserver/src/tenant/storage_layer/delta_layer.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (erikgrinaker) [reply pageserver/src/tenant/storage_layer/delta_layer.rs:1016]: > update_key shouldn't wait on the completion of the IO, so I don't really understand why we would make it async.

I think I just intuitively considered it easier to understand updating the state once the read completed, which we could do by nesting futures, but I don't think it has any practical consequence. We'd probably run into some ownership issues as well.

>I guess we could change VectoredValueReconstructState::on_disk_values to stash IO futures instead of oneshot channels, but I don't really see what that gets us.

I think there are a few benefits to consider:

* Spawning a task has a cost, and spawning many tasks puts pressure on the Tokio scheduler. I don't know if this matters considering we also do IO here, but it might at scale. Futures can be awaited on the current task without spawning new ones.

* Serial/parallel operation is trivial simply by awaiting the futures sequentially (serial) or using e.g. `FuturesOrdered` or `FuturesUnordered` to await them all concurrently on the same task (parallel) -- the latter also optimizes away polling costs.

* Futures are automatically cancelled when dropped, while spawned tasks need explicit cancellation (you alluded to this being a problem in the PR description).

* We avoid the additional channel synchronization and overhead. This probably doesn't matter considering we're doing IO.

[the remark this answers is not in the document]
Comment (VladLazar) [reply pageserver/src/tenant/storage_layer/delta_layer.rs:1016]: > I think I just intuitively considered it easier to understand updating the state once the read completed, which we could do by nesting futures, but I don't think it has any practical consequence. We'd probably run into some ownership issues as well.

Yeah, I agree. The issue with that is that we don't get concurrency. The gist here is that we move on to the next layer after issuing IO without waiting on completion.

> I think there are a few benefits to consider

That's fair. I'll give it a go when I get the time.

[the remark this answers is not in the document]
Comment (erikgrinaker) [reply pageserver/src/tenant/storage_layer/delta_layer.rs:1016]: > Yeah, I agree. The issue with that is that we don't get concurrency. The gist here is that we move on to the next layer after issuing IO without waiting on completion.

I meant that these would just wrap the IO future, but not await it straight away -- we'd await the futures concurrently. The state would update as each read completed. But as I said, this might cause ownership issues (we might get away with it if we're not spawning tasks, but could get ugly either way).

**Review by erikgrinaker — COMMENTED** [review]
I've done another pass to check that there aren't any issues with ordering or races, and I can't see any -- even though we dispatch IOs concurrently, we always access the results in a predetermined order.

I think this should be good to go, once we resolve the tasks vs. futures discussion above.

File: pageserver/src/tenant/storage_layer/delta_layer.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (VladLazar) [reply pageserver/src/tenant/storage_layer/delta_layer.rs:1016]: I gave this a go. It's doable but quite tricky. The difficulty comes from the fact that we need to fan out futures from the IO future itself (which reads data for multiple keys). This would be quite inefficient currently because we have to iterate over all `VectoredBlobsBuff::blobs` for each blob.

I think we should defer this to when we pick this back up and benchmark it. The task and future approaches should be benchmarked separately. With the task approach we actually perform IO while the traversal is happening, but with the futures approach all IO would be done at the end.

[the remark this answers is not in the document]
Comment (erikgrinaker) [reply pageserver/src/tenant/storage_layer/delta_layer.rs:1016]: Yeah, benchmarks seem reasonable. I'm a bit worried that putting this amount of pressure on the Tokio scheduler could cause latency issues, probably more than the gains we get from executing IOs earlier, but we can find out with appropriate benchmarks.

**Review by erikgrinaker — APPROVED** [review]
Discussed offline that we'll merge this as-is for testing and benchmarking, and revisit the futures vs tasks thing with benchmark data.

This change will still spawn tasks for each IO, but sequentially. That can affect latency, so maybe doublecheck that there's no regression in the disabled case.

File: pageserver/src/tenant/storage_layer/delta_layer.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (problame) [reply pageserver/src/tenant/storage_layer/delta_layer.rs:1020]: stale discussion

File: pageserver/src/tenant/storage_layer/delta_layer.rs [hunk 8]
```diff
@@ -998,94 +992,80 @@ impl DeltaLayerInner {
...
-
         let max_vectored_read_bytes = self
             .max_vectored_read_bytes
             .expect("Layer is loaded with max vectored bytes config")
             .0
             .into();
         let buf_size = Self::get_min_read_buffer_size(&reads, max_vectored_read_bytes);
-        let mut buf = Some(IoBufferMut::with_capacity(buf_size));
+
+        let mut ignore_key_with_err = None;
```
Comment (problame) [inline pageserver/src/tenant/storage_layer/delta_layer.rs:1002]: This value is `Copy`d into the spawned IO, but the code inside the spawned IO makes me think this was meant as a cancellation / "don't do more work because we failed an earlier IO" type thing.

File: .github/workflows/_build-and-test-locally.yml [hunk 9]
```diff
@@ -230,7 +230,7 @@ jobs:
 
           # run pageserver tests with different settings
           for io_engine in std-fs tokio-epoll-uring ; do
-            NEON_PAGESERVER_UNIT_TEST_VIRTUAL_FILE_IOENGINE=$io_engine ${cov_prefix} cargo nextest run $CARGO_FLAGS $CARGO_FEATURES  -E 'package(pageserver)'
+            NEON_PAGESERVER_VALUE_RECONSTRUCT_IO_CONCURRENCY=futures-unordered NEON_PAGESERVER_UNIT_TEST_VIRTUAL_FILE_IOENGINE=$io_engine ${cov_prefix} cargo nextest run $CARGO_FLAGS $CARGO_FEATURES  -E 'package(pageserver)'
```
Comment (problame) [inline .github/workflows/_build-and-test-locally.yml:233]: TO be revisited before merging

File: pageserver/src/tenant/storage_layer/delta_layer.rs (hunk 8 above)
Comment (VladLazar) [reply pageserver/src/tenant/storage_layer/delta_layer.rs:1006]: It was an optimisation to avoid doing extra work if we already failed a key. Looking at it now, it's of dubious value.
We save some CPU work, but that's it. We can either remove it altogether or make it more clear that it's per IO future by moving the initialization into the `for read ...` loop.

File: test_runner/performance/pageserver/test_page_service_batching.py [hunk 10]
```diff
@@ -31,46 +31,55 @@ class PageServicePipeliningConfigPipelined(PageServicePipeliningConfig):
...
 BATCHABLE: list[PageServicePipeliningConfig] = [PageServicePipeliningConfigSerial()]
-for max_batch_size in [1, 2, 4, 8, 16, 32]:
+for max_batch_size in [32]:
     for execution in EXECUTION:
         BATCHABLE.append(PageServicePipeliningConfigPipelined(max_batch_size, execution))
 
 
 @pytest.mark.parametrize(
-    "tablesize_mib, pipelining_config, target_runtime, effective_io_concurrency, readhead_buffer_size, name",
+    "tablesize_mib, pipelining_config, target_runtime, ps_io_concurrency, ps_direct_io_mode, effective_io_concurrency, readhead_buffer_size, name",
```
Comment (problame) [inline test_runner/performance/pageserver/test_page_service_batching.py:50]: remove the parametrization / stash it in a WIP PR

and simply rely on the global defaults for this PR, so that the nightly benches pick up the change

File: pageserver/src/tenant/timeline.rs [hunk 11]
```diff
@@ -1175,22 +1179,34 @@ impl Timeline {
...
         let layers_visited = reconstruct_state.get_layers_visited();
 
-        for (key, res) in std::mem::take(&mut reconstruct_state.keys) {
-            match res {
-                Err(err) => {
-                    results.insert(key, Err(err));
-                }
-                Ok(state) => {
-                    let state = ValueReconstructState::from(state);
+        let futs = FuturesUnordered::new();
```
Comment (problame) [inline pageserver/src/tenant/timeline.rs:1184]: Above, we bail if `get_vectored_reconstruct_data` returns an error while there are still IOs in flight. Need to fix this.

File: test_runner/performance/pageserver/test_page_service_batching.py [hunk 12]
```diff
@@ -11,7 +11,7 @@
 from fixtures.neon_fixtures import NeonEnvBuilder, PgBin, wait_for_last_flush_lsn
 from fixtures.utils import humantime_to_ms
 
-TARGET_RUNTIME = 30
+TARGET_RUNTIME = 5
```
[the remark this answers is not in the document]
Comment (problame) [reply test_runner/performance/pageserver/test_page_service_batching.py:14]: punted into https://github.com/neondatabase/neon/pull/10466

File: test_runner/performance/pageserver/pagebench/test_pageserver_max_throughput_getpage_at_latest_lsn.py [hunk 13]
```diff
@@ -56,13 +60,15 @@ def test_pageserver_characterize_throughput_with_n_tenants(
 # For reference, the space usage of the snapshots:
 # sudo du -hs /instance_store/neon/test_output/shared-snapshots/*
 # 19G	/instance_store/neon/test_output/shared-snapshots/max_throughput_latest_lsn-1-136
-@pytest.mark.parametrize("duration", [20 * 60])
+@pytest.mark.parametrize("duration", [20])
```
[the remark this answers is not in the document]
Comment (problame) [reply test_runner/performance/pageserver/pagebench/test_pageserver_max_throughput_getpage_at_latest_lsn.py:63]: punted into https://github.com/neondatabase/neon/pull/10466

File: test_runner/fixtures/pageserver/makelayers/l0stack.py [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+from fixtures.pageserver.utils import wait_for_last_record_lsn
+
+
+@dataclass
+class L0StackShape:
+    logical_table_size_mib: int = 50
+    delta_stack_height: int = 20
+
+
+def make_l0_stack(endpoint: Endpoint, shape: L0StackShape):
```
[the remark this answers is not in the document]
Comment (problame) [reply test_runner/fixtures/pageserver/makelayers/l0stack.py:18]: punted into https://github.com/neondatabase/neon/pull/10466

File: test_runner/deep_layers_with_delta.py [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+This showcases the need for concurrent IO.
+
+I'm not completely sure about it because I don't yet understand the relation between
+fillfactor and mvcc (does mvcc use the unfilled space in the page?).
+
+I think there's a chance that some of the deltas are will_init .
+Could be avoided with disabling fullpage writes.
+
+TODO: concurrent IO will read beyond will_init and discard the data (I think).
+Measure impact of that.
```
[the remark this answers is not in the document]
Comment (problame) [reply test_runner/deep_layers_with_delta.py:14]: punted into https://github.com/neondatabase/neon/pull/10466

File: pageserver/src/tenant/storage_layer.rs [hunk 16]
```diff
@@ -78,38 +86,106 @@ pub(crate) enum ValueReconstructSituation {
...
 impl VectoredValueReconstructState {
-    fn get_cached_lsn(&self) -> Option<Lsn> {
-        self.img.as_ref().map(|img| img.0)
-    }
-}
-
-impl From<VectoredValueReconstructState> for ValueReconstructState {
-    fn from(mut state: VectoredValueReconstructState) -> Self {
-        // walredo expects the records to be descending in terms of Lsn
-        state.records.sort_by_key(|(lsn, _)| Reverse(*lsn));
```
[the remark this answers is not in the document]
Comment (problame) [reply pageserver/src/tenant/storage_layer.rs:99]: TBH I never understood if this sort_by_key here had any effect.
The sequential code path was already putting in the records in the right order, right?

File: test_runner/performance/pageserver/test_page_service_batching.py (hunk 10 above)
Comment (problame) [reply test_runner/performance/pageserver/test_page_service_batching.py:50]: => https://github.com/neondatabase/neon/pull/10466

File: pageserver/src/tenant/storage_layer/delta_layer.rs (hunk 8 above)
Comment (problame) [reply pageserver/src/tenant/storage_layer/delta_layer.rs:1007]: I'll remove this.

Comment (problame) [reply pageserver/src/tenant/storage_layer/delta_layer.rs:1007]: I think the `ios.is_empty()` assert that we added would fire

File: pageserver/src/tenant/timeline.rs [hunk 17]
```diff
@@ -1175,22 +1179,34 @@ impl Timeline {
         let reconstruct_timer = crate::metrics::RECONSTRUCT_TIME
             .for_get_kind(get_kind)
             .start_timer();
```
Comment (problame) [inline pageserver/src/tenant/timeline.rs:1181]: This metric and the get_reconstruct_values data metric are incorrectly counted right now. They no longer have clear semantics in a concurrent world.

What to do about them?

File: pageserver/src/tenant/timeline.rs (hunk 11 above)
Comment (problame) [reply pageserver/src/tenant/timeline.rs:1184]: c4e0ba38b8b89807b177381803b0195c9d107aba

File: pageserver/src/tenant/storage_layer.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (problame) [reply pageserver/src/tenant/storage_layer.rs:99]: d6cdf1b13f7f81faac636fd4404edd36ccbb857b

File: pageserver/src/tenant/timeline.rs (hunk 17 above)
Comment (problame) [reply pageserver/src/tenant/timeline.rs:1192]: fe615520ddf3c024f6ef896cd2197017280b1dd5

File: pageserver/src/tenant/storage_layer.rs [hunk 18]
```diff
@@ -98,14 +98,65 @@ pub(crate) enum OnDiskValue {
...
+#[derive(Debug, thiserror::Error)]
+pub(crate) enum WaitCompletionError {
+    #[error("OnDiskValueIo was dropped without completing, likely the sidecar task panicked")]
+    IoDropped,
+}
+
+impl OnDiskValueIoWaiter {
+    pub(crate) async fn wait_completion(self) -> Result<OnDiskValueIoResult, WaitCompletionError> {
+        // NB: for Unnecessary IOs, this method never gets called because we don't add them to `on_disk_values`.
+        self.rx.await.map_err(|_| WaitCompletionError::IoDropped)
```
[the remark this answers is not in the document]
Comment (problame) [reply pageserver/src/tenant/storage_layer.rs:158]: This can totally happen, e.g., if the sidecar task panics.

**Review by Bodobolero — APPROVED** [review]
I verified that the two additional rust tests only add 1 minute to build time which should be ok.
